### PR TITLE
Create initial setup for testing

### PR DESCRIPTION
**Pull request by EchoAlice**

### CODE REVIEW

File: src/helper.rs [hunk 1]
```diff
@@ -16,3 +17,46 @@ pub struct Node {
...
+        ip_address: listen_addr,
+        udp_port: port_start,
+        node_id: [0_u8; 32],
+    };
+    our_nodes.push(local_node);
+    let first_node = Node {
+        ip_address: listen_addr,
+        udp_port: port_start + 1,
+        node_id: [1_u8; 32],
+    };
```
Comment (ralexstokes) [inline src/helper.rs]: a bit of a tangent but you could make a single helper fn to help you make these instances

something like:
```rust
fn mk_node(listen_addr: &Ipv4Addr, port_start: u16, index: usize) -> Node {
    Node {
         ip_address: listen_addr.clone(),
        udp_port: port_start + index as u16,
        node_id: [index as u8; 32],
}
```

then something like
```rust
let our_nodes: Vec<_> = (0..4).into_iter().map(|i| mk_node(listen_addr, port_start, i)).collect();
```

Comment (EchoAlice) [reply src/helper.rs]: Implemented here:  https://github.com/EchoAlice/my_kademlia/commit/5ebfa6fdb6f4bb5b29b2123e8f1912dc251e6dc2

File: src/helper.rs [hunk 2]
```diff
@@ -16,3 +17,35 @@ pub struct Node {
     pub udp_port: u16,
     pub node_id: Identifier,
 }
+
+pub mod testing {
+    use super::*;
+
+    pub fn mk_nodes() -> Vec<Node> {
+        // Should these nodes have different IP addresses?
```
Comment (ralexstokes) [inline src/helper.rs]: stylistic thing but I prefer to keep "thoughts" or questions like this out of the code; instead you could make an issue if its worthwhile or in case I'd think you can just keep a list of questions in a separate document locally

File: src/helper.rs [hunk 3]
```diff
@@ -16,3 +17,35 @@ pub struct Node {
...
+        let listen_addr = String::from("127.0.0.1").parse::<Ipv4Addr>().unwrap();
+        let port_start = 9000_u16;
+
+        let our_nodes: Vec<Node> = (0..5)
+            .into_iter()
+            .map(|i| mk_node(&listen_addr, port_start, i))
+            .collect();
+
+        our_nodes
```
Comment (ralexstokes) [inline src/helper.rs]: less code good

```suggestion
        (0..5)
            .into_iter()
            .map(|i| mk_node(&listen_addr, port_start, i))
            .collect()
```

File: src/helper.rs [hunk 4]
```diff
@@ -16,3 +17,35 @@ pub struct Node {
...
+            ip_address: listen_addr.clone(),
+            udp_port: port_start + index as u16,
+            node_id: [index as u8; 32],
+        }
+    }
+
+    pub fn mk_table(dummy_nodes: Vec<Node>) -> KbucketTable {
+        let listen_addr = String::from("127.0.0.1").parse::<Ipv4Addr>().unwrap();
+        let port_start = 9000_u16;
+
```
Comment (ralexstokes) [inline src/helper.rs]:
```suggestion
```

File: src/helper.rs [hunk 5]
```diff
@@ -16,3 +17,35 @@ pub struct Node {
...
+        KbucketTable::new(dummy_nodes[0].node_id)
```
Comment (ralexstokes) [inline src/helper.rs]: if this is all this fn is doing, you prolly can just delete it?

File: src/kbucket.rs [hunk 6]
```diff
@@ -115,34 +119,90 @@ impl KbucketTable {
...
-///
-/// Test find_node()      **Requires adding nodes to our table**
-/// Test search_table()
-/// Test add_node()
 #[cfg(test)]
 mod tests {
     use super::*;
+    use crate::helper::testing::{mk_nodes, mk_table};
+
+    // Run "cargo test add_node -- --nocapture" to see that node was added to routing table.
```
Comment (ralexstokes) [inline src/kbucket.rs]: this is fine to leave in but it is another type of comment you wouldn't see in most code imo

File: src/kbucket.rs [hunk 7]
```diff
@@ -115,34 +119,90 @@ impl KbucketTable {
...
 
-/// TODO:  Implement real deal tests!
-///
-/// Test find_node()      **Requires adding nodes to our table**
-/// Test search_table()
-/// Test add_node()
 #[cfg(test)]
 mod tests {
     use super::*;
+    use crate::helper::testing::{mk_nodes, mk_table};
```
Comment (ralexstokes) [inline src/kbucket.rs]: may be bike shedding but why not just define these few `fn`s you have here? rather than in `crate::helper::testing`?

File: src/kbucket.rs [hunk 8]
```diff
@@ -115,34 +119,90 @@ impl KbucketTable {
...
+
+    // Run "cargo test add_node -- --nocapture" to see that node was added to routing table.
+    #[test]
+    fn add_node() {
+        let dummy_nodes = mk_nodes();
+        let mut table = mk_table(dummy_nodes.clone());
```
Comment (ralexstokes) [inline src/kbucket.rs]: no reason to close all these nodes; I'd change the API here

```suggestion
        let mut table = mk_table(&dummy_nodes);
```

File: src/kbucket.rs [hunk 9]
```diff
@@ -115,34 +119,90 @@ impl KbucketTable {
...
+    #[test]
+    fn add_redundant_node() {
+        let dummy_nodes = mk_nodes();
+        let mut table = mk_table(dummy_nodes.clone());
+        let result = table.add_node(dummy_nodes[1]);
+        assert_eq!(true, result);
+        println!("Updated table: {:?}", table);
+        println!("\n");
+
+        let result = table.add_node(dummy_nodes[1]);
```
Comment (ralexstokes) [inline src/kbucket.rs]: and this?
```suggestion
        let result = table.add_node(dummy_nodes[1]);
        assert!(!result);
```

File: src/kbucket.rs [hunk 10]
```diff
@@ -115,34 +119,90 @@ impl KbucketTable {
...
+        assert_eq!(true, result);
+        println!("Updated table: {:?}", table);
+    }
+    #[test]
+    fn add_redundant_node() {
+        let dummy_nodes = mk_nodes();
+        let mut table = mk_table(dummy_nodes.clone());
+        let result = table.add_node(dummy_nodes[1]);
+        assert_eq!(true, result);
```
Comment (ralexstokes) [inline src/kbucket.rs]:
```suggestion
        assert!(result);
```

File: src/kbucket.rs [hunk 11]
```diff
@@ -115,34 +119,90 @@ impl KbucketTable {
...
+        let result = table.add_node(dummy_nodes[1]);
+    }
+
+    #[test]
+    fn search_table() {
+        let dummy_nodes = mk_nodes();
+        let mut table = mk_table(dummy_nodes.clone());
+        table.add_node(dummy_nodes[1]);
+
+        let result = table.search_table(dummy_nodes[1].node_id);
```
Comment (ralexstokes) [inline src/kbucket.rs]: ? check insertion result?

**Review by ralexstokes — COMMENTED** [review]
nice updates! left some comments/ questions

File: src/kbucket.rs (hunk 6 above)
Comment (EchoAlice) [reply src/kbucket.rs]: I'll put this in the justfile

File: src/kbucket.rs (hunk 7 above)
Comment (EchoAlice) [reply src/kbucket.rs]: I might place mk_nodes() in test mod.  Deleting mk_table.

File: src/kbucket.rs (hunk 8 above)
Comment (EchoAlice) [reply src/kbucket.rs]: Deleting this guy anyway

File: src/helper.rs (hunk 3 above)
Comment (EchoAlice) [reply src/helper.rs]: I ended up needing to use our_nodes here:  https://github.com/EchoAlice/my_kademlia/commit/c7aaf37239813e69b2f63df3de8dfef117b7fdad

Lmk what you think

File: src/kbucket.rs (hunk 8 above)
Comment (EchoAlice) [reply src/kbucket.rs]: But check out my return values for mk_nodes( ):  https://github.com/EchoAlice/my_kademlia/commit/c7aaf37239813e69b2f63df3de8dfef117b7fdad#diff-1a035035906977e355bc8da311ae54e7161dd36d85dd39bf81cbae9dabbf339aR149

What would you suggest doing?

File: justfile [hunk 12]
```diff
@@ -6,4 +6,7 @@ lint: fmt
     cargo clippy --all-targets
 build:
     cargo build --all-targets
-run-ci: lint build test
\ No newline at end of file
+run-ci: lint build test
+
+
+# Run "cargo test add_node -- --nocapture" to see that node was added to routing table.
```
Comment (ralexstokes) [inline justfile]:
```suggestion

run-test-with-logs TEST:
    cargo test {{TEST}} -- --nocapture
```

File: src/kbucket.rs [hunk 13]
```diff
@@ -115,34 +116,95 @@ impl KbucketTable {
...
+        let port_start = 9000_u16;
+
+        let our_nodes: Vec<Node> = (0..n)
+            .into_iter()
+            .map(|i| mk_node(&listen_addr, port_start, i))
+            .collect();
+
+        if let Some((local_node, other_nodes)) = our_nodes.split_first() {
+            let other_nodes = other_nodes.to_vec();
+            (local_node.clone(), other_nodes.clone())
```
Comment (ralexstokes) [inline src/kbucket.rs]:
```suggestion
            let other_nodes = other_nodes.to_vec();
            (local_node.clone(), other_nodes)
```

File: src/kbucket.rs [hunk 14]
```diff
@@ -115,34 +116,95 @@ impl KbucketTable {
...
+        Node {
+            ip_address: listen_addr.clone(),
+            udp_port: port_start + index as u16,
+            node_id: [index as u8; 32],
+        }
+    }
+
+    #[test]
+    fn add_redundant_node() {
+        let (local_node, dummy_nodes) = mk_nodes(2);
```
Comment (ralexstokes) [inline src/kbucket.rs]: this is more clear what you intend to happen imo
```suggestion
        let (local_node, remote_nodes) = mk_nodes(2);
```

File: src/kbucket.rs [hunk 15]
```diff
@@ -115,34 +116,95 @@ impl KbucketTable {
...
+        assert!(!result2);
+    }
+
+    #[test]
+    fn find_node_present() {
+        let (local_node, dummy_nodes) = mk_nodes(5);
+        let mut table = KbucketTable::new(local_node.node_id);
+        for i in 1..dummy_nodes.len() {
+            table.add_node(dummy_nodes[i]);
+        }
```
Comment (ralexstokes) [inline src/kbucket.rs]: I would just iter the `nodes` themselves

and generally I'd change `add_node` to take a `&Node`, rather than a `Node` directly

```suggestion
        for node in dummy_nodes {
            table.add_node(node);
        }
```

**Review by ralexstokes — COMMENTED** [review]
nice updates!

File: src/kbucket.rs (hunk 15 above)
Comment (EchoAlice) [reply src/kbucket.rs]: On it.  What's the reasoning behind passing a reference to .add_node() instead of passing a node directly?

Comment (ralexstokes) [reply src/kbucket.rs]: if the type is big enough it will be better to just pass a reference, rather than try to push through a bunch of data into the CPU caches

and as a rule of thumb if it is not clear from the type directly (e.g. u64 is clearly only 64 bits whereas I don't know about `Node` just from looking at it) then I just keep it simply and pass by reference where I can

File: src/helper.rs [hunk 16]
```diff
@@ -1,5 +1,6 @@
 #![allow(clippy::assign_op_pattern)]
 
+use crate::kbucket::KbucketTable;
```
Comment (ralexstokes) [inline src/helper.rs]: why is this here?

File: src/kbucket.rs [hunk 17]
```diff
@@ -115,34 +118,107 @@ impl KbucketTable {
...
+            let remote_nodes = remote_nodes.to_vec();
+            (*local_node, remote_nodes)
+        } else {
+            panic!("Not enough nodes were created");
+        }
+    }
+
+    fn mk_node(listen_addr: &Ipv4Addr, port_start: u16, index: u8) -> Node {
+        let mut node_id = [0_u8; 32];
+        node_id[31] += index;
```
Comment (ralexstokes) [inline src/kbucket.rs:150]: you don't need to change this but I try to avoid `mut` where possible

instead you could do something like `let node_id = [index; 32];`

File: src/kbucket.rs [hunk 18]
```diff
@@ -51,61 +54,61 @@ impl KbucketTable {
...
+    /// Instructs a node to store a key, value pair for later retrieval.
+    ///
+    /// "Most operations are implemented in terms of the lookup proceedure. To store a
+    /// <key,value> pair, a participant locates the k closes nodes to the key and sends them store RPCs".
     pub fn store(&mut self, key: Identifier, value: Vec<u8>) {}
 
     // Non-RPCs:
     // ---------------------------------------------------------------------------------------------------
-    pub fn add_node(&mut self, node: Node) {
+    // TODO:  Make add_node() take a reference to a node, not a node directly
```
Comment (ralexstokes) [inline src/kbucket.rs]:
```suggestion
```

File: src/kbucket.rs [hunk 19]
```diff
@@ -51,61 +54,61 @@ impl KbucketTable {
...
-
-        if !result.found {
-            bucket[result.column_index] = Some(node)
-        } else {
-            println!("Node is already in our table")
+        match result {
+            Search::Success(bucket_index, column_index) => false,
+            Search::Failure(bucket_index, column_index) => {
+                // Should I be dereferencing the node?  Or change buckets to take references to nodes?
+                self.buckets[bucket_index][column_index] = Some(*node);
```
Comment (ralexstokes) [inline src/kbucket.rs]: yes just deref ("copy") or even clone the node here

there are more sophisticated things you could end up doing but this is out of scope for now

File: src/kbucket.rs [hunk 20]
```diff
@@ -115,34 +118,107 @@ impl KbucketTable {
...
+                    if let Some(node) = node {
+                        let node_in_bucket_index = table.xor_bucket_index(node.node_id);
+                        assert_ne!(node_to_find, node);
+                        assert_eq!(node_to_find_index, node_in_bucket_index);
+                    } else {
+                        panic!("find_node() returned an empty index")
+                    }
+                }
+            }
+            _ => panic!("FindNodeResult shouldn't == Found"),
```
Comment (ralexstokes) [inline src/kbucket.rs]: no need to change but FYI: https://doc.rust-lang.org/std/macro.unreachable.html

**Review by ralexstokes — APPROVED** [review]
looks good; I still am confused on the relation of `add_node` and `find_node` but I think it will become clearer as you make more progress

File: src/kbucket.rs (hunk 17 above)
Comment (EchoAlice) [reply src/kbucket.rs:150]: Ahhhh...

I changed node_id from [index; 32] to this mut node_id so it's easier for me to reason about bucket distances.
